### ISSUE TITLE
Hanging when formatting a decently sized math expression

### ISSUE THREAD

**Issue by PvdBerg1998** [rust-lang/rustfmt#5861]
I am porting some ancient Fortran code, which contains the following mess:

```rust
q1[n - 1] = d0*(j1*(phi/2.0+v*ds)).exp() / (
    1.0 + v.powi(2) / (
        w.powi(2) +
        (
            d0 / (1.0 + v.powi(2) / (
                w.powi(2) + (d0 / (1.0 + v.powi(2) / (
                    w.powi(2) + (d0 / (1.0 + v.powi(2) / (
                        w.powi(2) + (d0 / (1.0 + v.powi(2) / (
                            w.powi(2) + (d0 / (1.0 + v.powi(2) / (
                                w.powi(2) + d0.powi(2)
                            ).sqrt())).powi(2)
                        ).sqrt())).powi(2)
                    ).sqrt())).powi(2)
                ).sqrt())).powi(2)
        ).sqrt())).powi(2)
    ).sqrt()
);
```

When attempting to format this code, rustfmt appears to run indefinitely.

**Comment by xxchan** [comment]
reproduceable example
```rust

fn main() {
    let mut q1: Vec<f64> = vec![0.0; 100];
    let n: usize =1;
    let d0: f64 = 1.0;
    let j1: f64 = 1.0;
    let phi: f64 = 1.0;
    let v: f64 = 1.0;
    let w: f64 = 1.0;
    let ds: f64 = 1.0;


    q1[n - 1] = d0*(j1*(phi/2.0+v*ds)).exp() / (
        1.0 + v.powi(2) / (
            w.powi(2) +
            (
                d0 / (1.0 + v.powi(2) / (
                    w.powi(2) + (d0 / (1.0 + v.powi(2) / (
                        w.powi(2) + (d0 / (1.0 + v.powi(2) / (
                            w.powi(2) + (d0 / (1.0 + v.powi(2) / (
                                w.powi(2) + (d0 / (1.0 + v.powi(2) / (
                                    w.powi(2) + d0.powi(2)
                                ).sqrt())).powi(2)
                            ).sqrt())).powi(2)
                        ).sqrt())).powi(2)
                    ).sqrt())).powi(2)
            ).sqrt())).powi(2)
        ).sqrt()
    );
}

```

**Comment by calebcartwright** [comment]
Thanks for the report.

TL;DR - I don't think this is going to change any time soon, so you may want to consider refactoring to usage of more locals if feasible, or using the `#[rustfmt::skip]` attribute (which may also require some minor refactoring depending on whether you're using a nightly toolchain)

As to the likely "why":

Given how this snippet would be represented in the AST (at least based on a quick mental parse), I suspect this is a storm of pathological input (for rustfmt :smile:) that probably hits on several of our ugliest, worst case scenario paths in the rustfmt source; namely, the complex, nested chains and bin expressions that's even more exacerbated by appearing in a RHS position.

There's a lot of unnecessary allocations triggered along those paths, and extensive "retries" to identify a preferable format which, IIRC, was found to have a cubic time complexity in the worst case.

We've really only seen reports hitting those cases with large, generated code and although attempts have been made to improve things, there's some friction trying to get the algorithmic complexity under control without negatively impacting the quality of formatting in more normal/less complex input cases.

**Comment by knipknap** [comment]
Another severe instance of this — formatting a degree-11 Horner-form polynomial (two copies, ~50 lines total) takes **~70 seconds** on rustfmt 1.9.0 (rustc 1.95.0).

Reproduction:

```rust
fn fast_acos(x: f64) -> f64 {
    const C0: f64 = 1.0000000000000000e+00;
    const C1: f64 = 1.6666666666689003e-01;
    const C2: f64 = 7.4999999956843033e-02;
    const C3: f64 = 4.4642860300686517e-02;
    const C4: f64 = 3.0381825038220060e-02;
    const C5: f64 = 2.2374824118139486e-02;
    const C6: f64 = 1.7315158262635719e-02;
    const C7: f64 = 1.4311872250111499e-02;
    const C8: f64 = 9.4415958559379339e-03;
    const C9: f64 = 1.8048381466995028e-02;
    const C10: f64 = -1.1324721050537142e-02;
    const C11: f64 = 3.1226116192186414e-02;

    let t = x * x;
    let p = C0
        + t * (C1
            + t * (C2
                + t * (C3
                    + t * (C4
                        + t * (C5
                            + t * (C6
                                + t * (C7
                                    + t * (C8
                                        + t * (C9
                                            + t * (C10 + t * C11))))))))));
    std::f64::consts::FRAC_PI_2 - x * p
}
```

`time cargo fmt` on just this one function: ~29s with a single copy, ~70s with two copies (as in the original).
